consul/src/main/java/com/linecorp/armeria/server/consul/ConsulUpdatingListener.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.consul;
+
+import static java.util.Objects.requireNonNull;
+
+import java.net.Inet4Address;
+import java.net.URI;
+import java.util.concurrent.ThreadLocalRandom;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.util.SystemInfo;
+import com.linecorp.armeria.internal.consul.Check;
+import com.linecorp.armeria.internal.consul.ConsulClient;
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerListener;
+import com.linecorp.armeria.server.ServerListenerAdapter;
+import com.linecorp.armeria.server.ServerPort;
+
+/**
+ * A {@link ServerListener} which registers the current {@link Server} to
+ * <a href="https://www.consul.io">Consul</a>.
+ */
+public class ConsulUpdatingListener extends ServerListenerAdapter {
+
+    private static final Logger logger = LoggerFactory.getLogger(ConsulUpdatingListener.class);
+
+    /**
+     * Returns a {@link ConsulUpdatingListenerBuilder} that builds {@link ConsulUpdatingListener}.
+     * @param serviceName the service name which is registered into Consul.
+     */
+    public static ConsulUpdatingListenerBuilder builder(String serviceName) {
+        return new ConsulUpdatingListenerBuilder(serviceName);
+    }
+
+    private final ConsulClient consulClient;
+    private final String serviceName;
+
+    @Nullable
+    private final Endpoint endpoint;
+    @Nullable
+    private final Check check;
+    @Nullable
+    private String serviceId;
+
+    ConsulUpdatingListener(ConsulClient consulClient, String serviceName, @Nullable Endpoint endpoint,
+                           @Nullable URI checkUrl, @Nullable HttpMethod checkMethod, String checkInterval) {
+        this.consulClient = requireNonNull(consulClient, "consulClient");
+        this.serviceName = requireNonNull(serviceName, "serviceName");
+        this.endpoint = endpoint;
+
+        if (checkUrl != null) {
+            final Check check = new Check();
+            check.setHttp(checkUrl.toString());
+            if (checkMethod != null) {
+                check.setMethod(checkMethod.toString());
```

Comment:
`.name()` because `toString()` is for human-friendly representation?

Reply:
Yes.